### ISSUE TITLE
Object created by omitting a default value is not equal to object created with explicitly setting the default value

### ISSUE THREAD

**Issue by martina-oefelein** [pydantic/pydantic#5432]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

If a field has a default value, omitting the field when creating the object should have the same effect as explicitly setting the default value.

The objects look the same, but when compared for equality, this returns false.

The [announcement](https://docs.pydantic.dev/blog/pydantic-v2-alpha/#changes-to-basemodel) says:

> The `__eq__` method has changed for models; models are no longer considered equal to the dicts.

I wonder whether it's related to this change?

### Example Code

```Python
from pydantic import BaseModel


class SimpleModel(BaseModel):
    """Model with some simple fields."""

    field: str = "Test"


def test_default_value() -> None:
    """Omitting a value should be equivalent to setting it to its default.

    This works in Pydantic 1, but fails in Pydantic 2.
    """
    assert SimpleModel() == SimpleModel(field="Test")
    # E       AssertionError: assert SimpleModel(field='Test') == SimpleModel(field='Test')
    # E         Full diff:
    # E           SimpleModel(field='Test')
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0a1
                 install path: /Users/moefelei/Repos/cli-prototype/.venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.2 (main, Feb 16 2023, 03:07:35) [Clang 14.0.0 (clang-1400.0.29.202)]
                     platform: macOS-13.2.1-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

**Comment by dmontagu** [comment]
I definitely agree this would be a significant bug, but I can't seem to reproduce it:
```
from pydantic import BaseModel


class SimpleModel(BaseModel):
    """Model with some simple fields."""

    field: str = "Test"


def test_default_value() -> None:
    """Omitting a value should be equivalent to setting it to its default.

    This works in Pydantic 1, but fails in Pydantic 2.
    """
    assert SimpleModel() == SimpleModel(field="Test")
    print("success")


test_default_value()
#> success
```
Here's the relevant pieces I get from running `pip freeze`:
```
❯ pip freeze | grep pydantic         
pydantic==2.0a1
pydantic_core==0.23.0
```
I tested it on both python 3.10 and python 3.11.

**Comment by martina-oefelein** [comment]
I've never had pydantic installed from a clone of the repo.
I'm using poetry, and switch between pydantic versions by updating the version in pyproject.toml and running `poetry lock --no-update && poetry install`
I re-created the venv, same result.

I also tried to reproduce it using new environment with just pydantic 2.0a1 in it, same result.

I noticed you have pydantic_core==0.23.0, while I have 0.23.1, but downgrading to 0.23.0 didn't change anything

Anything else I can try?

**Comment by martina-oefelein** [comment]
Also tried it with Python 3.10. Same result

```
             pydantic version: 2.0a1
                 install path: /Users/moefelei/Repos/pydantic-sandbox/.venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.10 (main, Feb 16 2023, 02:55:02) [Clang 14.0.0 (clang-1400.0.29.202)]
                     platform: macOS-13.2.1-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

**Comment by dmontagu** [comment]
@martina-oefelein I've successfully reproduced your issue. I will see if I can fix it.

**Comment by dmontagu** [comment]
@martina-oefelein For what it's worth, I don't get the error with `pydantic_core==0.22.0`. So you should be able to install that version. I'll try to get it so that using the latest compatible versions of `pydantic_core` and `pydantic` together doesn't have this bug ASAP

**Comment by dmontagu** [comment]
@martina-oefelein in the just-released `pydantic==2.0a2`, we have pinned the version of pydantic-core, so hopefully issues like this won't happen any more. Please let us know as you find more issues, and thank you for continuing to report the ones you have been finding!

**Comment by hramezani** [comment]
As we haven't received any feedback we consider this issue as fixed.
Feel free to open the issue again or create a new one in case of problem.
